transformer_engine/common/util/system.h

```diff
@@ -7,25 +7,96 @@
 #ifndef TRANSFORMER_ENGINE_COMMON_UTIL_SYSTEM_H_
 #define TRANSFORMER_ENGINE_COMMON_UTIL_SYSTEM_H_
 
+#include <cstdint>
+#include <cstdlib>
+#include <filesystem>
+#include <fstream>
+#include <sstream>
 #include <string>
 
+#include "logging.h"
+
 namespace transformer_engine {
 
+namespace detail {
+
+/*! \brief Template specialization to get the env var for numeric data types */
+template <typename T>
+inline typename std::enable_if<std::is_arithmetic<T>::value, T>::type getenv_helper(
+    const char *variable, const T &default_value) {
+  // Implementation for numeric types
+  const char *env = std::getenv(variable);
+  if (env == nullptr || env[0] == '\0') {
+    return default_value;
+  }
+  T value;
+  std::istringstream iss(env);
+  iss >> value;
+  NVTE_CHECK(iss, "Invalid environment variable value");
+  return value;
+}
+
+/*! \brief Template specialization to get the env var for string-like data types */
+template <typename T>
+inline typename std::enable_if<!std::is_arithmetic<T>::value, T>::type getenv_helper(
+    const char *variable, const T &default_value) {
+  // Implementation for string-like types
+  const char *env = std::getenv(variable);
+  if (env == nullptr || env[0] == '\0') {
+    return default_value;
+  } else {
+    return env;
+  }
+}
+
+/*! \brief Template specialization to get the default values for different
+* numeric data types
+*/
+template <typename T>
+inline T getenv_default_value() {
+  return 0;
+}
+
+/*! \brief Template specialization to get the default values for bool */
+template <>
+inline bool getenv_default_value<bool>() {
+  return false;
+}
+
+/*! \brief Template specialization to get the default values for string */
+template <>
+inline std::string getenv_default_value<std::string>() {
+  return std::string();
+}
+
+/*! \brief Template specialization to get the default values for filesystem
+* path data type */
+template <>
+inline std::filesystem::path getenv_default_value<std::filesystem::path>() {
+  return std::filesystem::path();
+}
+
+}  // namespace detail
+
 /*! \brief Get environment variable and convert to type
  *
- * If the environment variable is unset or empty, a falsy value is
+ * If the environment variable is unset or empty, a false value is
```

Comment:
Nit: "falsy" is more accurate since we're not actually returning `false` (https://developer.mozilla.org/en-US/docs/Glossary/Falsy). That said, it seems "falsy" and "truthy" are more used in the JavaScript world.
```suggestion
 * If the environment variable is unset or empty, a falsy value is
```

Reply:
Thanks for the link to falsy. I wasn't aware of falsy and assumed it was a typo :P

Reply:
Seems like it's slang. When looking for the formal definition, I was surprised that it isn't really used anywhere official except JavaScript.